/*Copyright 2024 Ross Video Ltd
*
* Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
*
* 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
*
* 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
*
* 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES, 
* INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, 
* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. */


syntax = "proto3";

package catena;

import "language.proto";

option java_multiple_files = true;
option java_package = "catena.core.menu";
option java_outer_classname = "MenuProtos";

option optimize_for = SPEED;

/* Menu message.
 * Defines a logical grouping of commands and parameters that can be displayed
 * together in a GUI and have client access control applied.
 * Note that sub-parameters (and sub-sub-parameters) can be selected by
 * specifying a JSON pointer relative to the device's "params" component. */
message Menu {
  /* Display name for the Menu */
  PolyglotText name = 1;

  /* Whether the menu should be hidden in the client GUI. */
  bool hidden = 2;

  /* Whether the menu should be disabled (shown as read-only) in the client GUI. */
  bool disabled = 3;

  /* The menu's parameter members */
  repeated string param_oids = 4;

  /* The menu's command members */
  repeated string command_oids = 5;

  /* Locator for a device specific custom user interface descriptor.
   * Additional properties to assist client including ui-url */
  map<string, string> client_hints = 6;
}

/* A group of 'theme' related menus. */
message MenuGroup {
  /* The name of the menu group for display in client GUI */
  PolyglotText name = 1;

  /* The menus in the group */
  map<string, Menu> menus = 2;
}
